streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala

```diff
@@ -170,101 +207,64 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
-      s"$message - $error"
-    } else {
-      s"$message"
-    }
-    logError(s"Deregistered receiver for stream $streamId: $messageWithError")
+  /**
+   * Get the list of executors excluding driver
+   */
+  private def getExecutors(ssc: StreamingContext): List[String] = {
+    val executors = ssc.sparkContext.getExecutorMemoryStatus.map(_._1.split(":")(0)).toList
+    val driver = ssc.sparkContext.getConf.get("spark.driver.host")
+    executors.diff(List(driver))
   }
 
-  /** Add new blocks for the given stream */
-  private def addBlock(receivedBlockInfo: ReceivedBlockInfo): Boolean = {
-    receivedBlockTracker.addBlock(receivedBlockInfo)
-  }
+  /** Check if tracker has been marked for starting */
+  private def isTrackerStarted(): Boolean = trackerState == Started
 
-  /** Report error sent by a receiver */
-  private def reportError(streamId: Int, message: String, error: String) {
-    val newReceiverInfo = receiverInfo.get(streamId) match {
-      case Some(oldInfo) =>
-        oldInfo.copy(lastErrorMessage = message, lastError = error)
-      case None =>
-        logWarning("No prior receiver info")
-        ReceiverInfo(streamId, "", null, false, "", lastErrorMessage = message,
-          lastError = error, lastErrorTime = ssc.scheduler.clock.getTimeMillis())
-    }
-    receiverInfo(streamId) = newReceiverInfo
-    listenerBus.post(StreamingListenerReceiverError(receiverInfo(streamId)))
-    val messageWithError = if (error != null && !error.isEmpty) {
-      s"$message - $error"
-    } else {
-      s"$message"
-    }
-    logWarning(s"Error reported by receiver for stream $streamId: $messageWithError")
-  }
+  /** Check if tracker has been marked for stopping */
+  private def isTrackerStopping(): Boolean = trackerState == Stopping
 
-  /** Check if any blocks are left to be processed */
-  def hasUnallocatedBlocks: Boolean = {
-    receivedBlockTracker.hasUnallocatedReceivedBlocks
-  }
+  /** Check if tracker has been marked for stopped */
+  private def isTrackerStopped(): Boolean = trackerState == Stopped
 
   /** RpcEndpoint to receive messages from the receivers. */
   private class ReceiverTrackerEndpoint(override val rpcEnv: RpcEnv) extends ThreadSafeRpcEndpoint {
 
+    /**
+     * Track all receivers' information. The key is the receiver id, the value is the receiver info.
+     */
+    private val receiverTrackingInfos = new HashMap[Int, ReceiverTrackingInfo]
+
+    /**
+     * Store all preferred locations for all receivers. We need this information to schedule
+     * receivers
+     */
+    private val receiverPreferredLocations = new HashMap[Int, Option[String]]
+
+    // TODO Remove this thread pool after https://github.com/apache/spark/issues/7385 is merged
+    private val submitJobThreadPool = ExecutionContext.fromExecutorService(
+      ThreadUtils.newDaemonCachedThreadPool("submit-job-thead-pool"))
+
     override def receive: PartialFunction[Any, Unit] = {
```

Comment:
Actually, `receiverAndReply` needs to create a temp `Actor` to send the message. Since we don't need to reply anything, it's better to use `receive`.